kaska/smoothn.py

```diff
@@ -1,537 +1,617 @@
-            + 'has been reached. Put s as an input variable if required.'])
-    elif abs(log10(s)-log10(sMaxBnd))<errp:
-        warning('MATLAB:smoothn:SUpperBound',\
-            ['s = %.3f '%(s) + ': the upper bound for s '\
-            + 'has been reached. Put s as an input variable if required.'])
-    #warning('MATLAB:smoothn:MaxIter',\
-    #    ['Maximum number of iterations (%d'%(MaxIter) + ') has '\
-    #    + 'been exceeded. Increase MaxIter option or decrease TolZ value.'])
-  return z,s,exitflag,Wtot
-
-def warning(s1,s2):
-  print(s1)
-  print(s2[0])
-
-## GCV score
-#---
-#function GCVscore = gcv(p)
-def gcv(p,Lambda,aow,DCTy,IsFinite,Wtot,y,nof,noe,smoothOrder):
-    # Search the smoothing parameter s that minimizes the GCV score
-    #---
-    s = 10**p;
-    Gamma = 1./(1+(s*abs(Lambda))**smoothOrder);
-    #--- RSS = Residual sum-of-squares
-    if aow>0.9: # aow = 1 means that all of the data are equally weighted
+        return np.array([np.log10(s)])
+
+
+#  GCV score
+# ---
+# function GCVscore = gcv(p)
+def gcv(p, lambda_v, aow, dct_y, is_finite, w_tot, y, nof, noe, smooth_order):
+    """
+    Search the smoothing parameter s that minimizes the GCV score
+    """
+    s = 10 ** p
+    gamma = 1.0 / (1 + (s * np.abs(lambda_v)) ** smooth_order)
+    # --- rss = Residual sum-of-squares
+    if aow > 0.9:  # aow = 1 means that all of the data are equally weighted
         # very much faster: does not require any inverse DCT
-        RSS = norm(DCTy*(Gamma-1.))**2;
+        rss = norm(dct_y * (gamma - 1.0)) ** 2
     else:
-        # take account of the weights to calculate RSS:
-        yhat = dctND(Gamma*DCTy,f=idct);
-        RSS = norm(sqrt(Wtot[IsFinite])*(y[IsFinite]-yhat[IsFinite]))**2;
-    #---
-    TrH = sum(Gamma);
-    GCVscore = RSS/float(nof)/(1.-TrH/float(noe))**2;
-    return GCVscore
-
-## Robust weights
-#function W = RobustWeights(r,I,h,wstr)
-def RobustWeights(r,I,h,wstr):
-    # weights for robust smoothing.
-    MAD = median(abs(r[I]-median(r[I]))); # median absolute deviation
-    u = abs(r/(1.4826*MAD)/sqrt(1-h)); # studentized residuals
-    if wstr == 'cauchy':
-        c = 2.385; W = 1./(1+(u/c)**2); # Cauchy weights
-    elif wstr == 'talworth':
-        c = 2.795; W = u<c; # Talworth weights
+        # take account of the weights to calculate rss:
+        yhat = dctND(gamma*dct_y, f=idct)
+        rss = norm(np.sqrt(w_tot[is_finite]) *
+                   (y[is_finite] - yhat[is_finite])) ** 2
+    # ---
+    tr_h = np.sum(gamma)
+    gcv_score = rss / float(nof) / (1.0 - tr_h / float(noe)) ** 2
+    return gcv_score
+
+
+#  Robust weights
+# function W = robust_weights(r,I,h,wstr)
+def robust_weights(r, i, h, wstr):
+    """
+    Recalculate the weights for robust smoothing.
+    """
+    mad = np.median(np.abs(r[i] - np.median(r[i])))  # median absolute deviation
+    u = np.abs(r / (1.4826 * mad) / np.sqrt(1 - h))  # studentized residuals
```

Comment:
Maybe provide some references for the hard-coded numbers? (@jgomezdans ?)

Reply:
Those appear in the original paper. They're constants that have been pre-calculated. To be honest, I don't see anything wrong with having some of those numbers around. I know that they're often used in robust statistics, so I don't really see any problem with them. Basically, if you haven't read the paper, you shouldn't be touching the code.

Reply:
That's fine, I just meant putting the references next to the numbers, like Tim did in his last commit.